File: uservice-reports/main/src/main/scala/sinnet/Mapper.scala
package sinnet

import sinnet.reports.report1.{ReportRequest => ReportRequestDTO, ReportRequests => ReportRequestsDTO}
import java.time.LocalDate
import scala.collection.JavaConverters._
import scala.util.Try

/** Converts ReportRequest (DTO) to its local model. */
object Mapper {
  private implicit def toLocalDate(x: sinnet.reports.Date): Option[LocalDate] = 
    Try(LocalDate.of(x.getYear(), x.getMonth(), x.getDayOfTheMonth())).toOption
  private implicit def toActivityDetails(x: sinnet.reports.report1.ActivityDetails) =
    ActivityDetails(x.getDescription(), x.getWho(), x.getWhen(), new Minutes(x.getHowLongInMins()), new Kilometers(x.getHowFarInKms()))
  private implicit def toCustomerDetails(x: sinnet.reports.report1.CustomerDetails) =
    CustomerDetails(x.getCustomerName(), x.getCustomerCity(), x.getCustomerAddress())

    
  implicit def apply(dto: ReportRequestDTO): ReportRequest = {
    ReportRequest(dto.getCustomer(), dto.getDetailsList().asScala.map(toActivityDetails _))
  }
}

File: uservice-timeentries/main/host/src/main/java/sinnet/reports/report1/Mapper.java
package sinnet.reports.report1;

import io.vavr.collection.List;
import io.vavr.control.Option;
import sinnet.reports.Kilometers;
import sinnet.reports.Minutes;

/** Converts ReportRequest (DTO) to its local model. */
class Mapper {

  private static ActivityDetails fromDto(sinnet.report1.grpc.ActivityDetails x) {
    var maybeWhen = Option.of(x.getWhen()).filter(it -> x.hasWhen()).flatMap(it -> sinnet.reports.Mapper.fromDto(it));
    return new ActivityDetails(x.getDescription(), x.getWho(), maybeWhen, Minutes.of(x.getHowLongInMins()),
        Kilometers.of(x.getHowFarInKms()));
  }

  private static CustomerDetails fromDto(sinnet.report1.grpc.CustomerDetails x) {
    return new CustomerDetails(x.getCustomerName(), x.getCustomerCity(), x.getCustomerAddress());
  }

  public static ReportRequest fromDto(sinnet.report1.grpc.ReportRequest dto) {
    var items = List.ofAll(dto.getDetailsList()).map(Mapper::fromDto);
    var customer = fromDto(dto.getCustomer());
    return new ReportRequest(customer, items);
  }

}
